src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    /// <summary>
+    /// This is a base class for LINQ related conversions between LINQ queries, LINQ methods and foreach loops.
+    /// </summary>
+    internal abstract class AbstractConvertLinqProvider : CodeRefactoringProvider
+    {
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
+
+            await CreateAnalyzer(semanticModel, context.CancellationToken)
+                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
+        }
+
+        protected abstract IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        protected interface IAnalyzer
+        {
+            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
+        }
+
+        protected abstract class AnalyzerBase<TRefactor, TSource, TDestination> : IAnalyzer
+            where TRefactor : SyntaxNode
+            where TSource : SyntaxNode
+            where TDestination : SyntaxNode
+        {
+            protected readonly SemanticModel _semanticModel;
+            protected readonly CancellationToken _cancellationToken;
+
+            public AnalyzerBase(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var nodeToRefactor = FindNodeToRefactor(root, context);
+                if (nodeToRefactor == null)
+                {
+                    return;
+                }
+
+                if (nodeToRefactor.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                if (TryConvert(nodeToRefactor, out var result))
```

Comment:
>TryConvert(nodeToRefactor, out var result [](start = 20, length = 41)

I presume the `DocumentUpdate` result contains some analysis info that is passed onto the eventual code action? Can we please add some doc comments on `DocumentUpdate` to describe it better?

nit: probably suffix it with `Info`?